cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -165,10 +168,19 @@ SQLRETURN SQLGetDiagFieldW(SQLSMALLINT handleType, SQLHANDLE handle,
 
     case SQL_HANDLE_DBC: {
       ODBCConnection* connection = reinterpret_cast<ODBCConnection*>(handle);
+      dsn = connection->GetDSN();
```

Comment:
This copies the DSN string even if SQL_DIAG_SERVER_NAME isn't being used. Can this happen at the field? Will need to switch on HandleType again.

Also, SQL_DIAG_SERVER_NAME should report the DSN for handle types other than connection. When statement and descriptor support is added, they should try to get it from the owning connection. (add a TODO to do this so it's not missed).